### ISSUE TITLE
Nested list in Table of Contents

### ISSUE THREAD

**Issue by tingyuw** [JacquesCarette/Drasil#2915]
The current way to build the table of contents is by passing the ``DocDesc``, 
```
toToC :: DocSection -> ItemType
toToC TableOfContents      = mktToCSec
toToC (RefSec rs)          = mktRefSec rs
toToC (IntroSec i)         = mktIntroSec i
toToC (StkhldrSec sts)     = mktStkhldrSec sts
toToC (GSDSec gs')         = mktGSDSec gs'
toToC (SSDSec ss)          = mktSSDSec ss
toToC (ReqrmntSec r)       = mktReqrmntSec r
toToC (LCsSec lc)          = mktLCsSec lc
toToC (UCsSec ulcs)        = mktUCsSec ulcs
toToC (TraceabilitySec t)  = mktTraceabilitySec t
toToC (AuxConstntSec acs)  = mktAuxConsSec acs
toToC Bibliography         = mktBib
toToC (AppndxSec a)        = mktAppndxSec a
toToC (OffShelfSolnsSec o) = mktOffShelfSolnSec o
```
and since the subsections are contained in the main section, we can easily map the subsections to build a nested list. 
```
-- | Helper for creating the 'Reference Material' section ToC entry
mktRefSec :: RefSec -> ItemType
mktRefSec (RefProg _ l) =
  mkHeaderItem (namedRef SRS.refMatLabel $ titleize Doc.refMat) $ map mktSubRef l
  where
    mktSubRef :: RefTab -> Sentence
    mktSubRef TUnits        = namedRef SRS.tOfUnitLabel   $ titleize' Doc.tOfUnit
    mktSubRef (TUnits' _ _) = namedRef SRS.tOfUnitLabel   $ titleize' Doc.tOfUnit
    mktSubRef (TSymb _)     = namedRef SRS.tOfSymbLabel   $ titleize' Doc.tOfSymb
    mktSubRef (TSymb' _ _)  = namedRef SRS.tOfSymbLabel   $ titleize' Doc.tOfSymb
    mktSubRef TAandA        = namedRef SRS.tOfAbbAccLabel $ titleize' Doc.abbAcc
```
But the nested sections are removed, each (sub)section is an independent ``DocSection``, I'm not sure how to map the relative subsections under their main section. I remember last time I brought up this issue, the solution that brings up is to have two section representations, one as the inner structure (non-nested one), and the other nested one for better display. Should I proceed with this solution? Or is there a better approach to rebuild this section without having two representations?

**Comment by JacquesCarette** [comment]
I'm not sure I fully understand, so what I propose here might not fit.

I think that all sections should have a 'parent' field, indicating what it is a sub-section of. There should be a special 'Top' section, that is the parent of normal sections.

**Comment by tingyuw** [comment]
I think we do need to indicate the parent of all sections, but this might be another designs. The first thing might be adding a ``UID`` in the section representation?

I guess my main question about this issue is how do we make a flatten list nested? For the table of contents, we need to find all possible sections and subsections to make this section. And we achieve that by passing the ``SRSDecl`` (i.e., ``[DocSection]``) so that we know what sections are included in this particular example. But now all sections are listed (flatten) in the ``[DocSection]`` like this
```
data DocSection = TableOfContents                      
                | RefSec DL.RefSec                     
                | TUnits DL.TUnits
                | TUnits' DL.TUnits'
                | TSymb DL.TSymb
                | TSymb' DL.TSymb'
                | TAandA DL.TAandA
                | IntroSec DL.IntroSec                
                | IPurposeSub DL.IPurposeSub
                | IScopeSub DL.IScopeSub
                | ICharSub DL.ICharSub
                | IOrgSub DL.IOrgSub
                | StkhldrSec DL.StkhldrSec             
                | ClientSub DL.ClientSub 
                | CstmrSub DL.CstmrSub
                | GSDSec DL.GSDSec            
                | SysCntxt DL.SysCntxt
                | UsrChars DL.UsrChars
                | SystCons DL.SystCons
...
```
how do we know which sections to build by the ``SRSDecl`` and also map the relative subsections together under their parent?

**Comment by JacquesCarette** [comment]
Yes, adding parents by `UID` is probably a good idea.

A quick hack for nesting would be to build a Rose Tree (of DocSection) as the input. The tree would indicate depth and parent, all at once. This tree would be explicitly built in each example.

**Comment by tingyuw** [comment]
Oh, it seems to be feasible. So it would be like first creating a tree data structure in the ``drasil-lang`` (maybe?) and then build the tree explicitly in each example. In that case, we will have two section structures for each example and the table of contents section would be built separately because it takes different input from the other sections right?

**Comment by JacquesCarette** [comment]
Yes to first creating a tree data structure. You might as well use https://hackage.haskell.org/package/rose-trees-0.0.4.5 instead of creating a new one.

Not sure I understand the "in that case", but let me explain my understanding:
- the sections (and subsections) will be build separately
- they will be assembled into a document using the rose tree structure
- the table of contents machinery will have to use that instead.

**Comment by tingyuw** [comment]
I add a rose tree data structure in ``Langauge/Drasil/Document.hs``. 
```data SecTree a = Node a [SecTree a]```

And I was thinking to build a ``SecTree DocSection`` in each example, which will look like
```
toC = 
  Node TableOfContents [
    Node (RefSec $ RefProg intro) [
      Node (TUnits TUProg) [], 
      Node (TSymb $ tsymb [TSPurpose, SymbOrder]) [], 
      Node (TAandA TAAProg) []],
    ...]
```
But we will have to build the sections twice, one is the original flatten structure to build each section except table of contents section (i.e. ``mkSRS``), the other is the recursive tree structure, would that be okay?

**Comment by JacquesCarette** [comment]
Why not reuse the one from the package I linked? That would give you quite a few functions "for free" at the same time.

Furthermore, you should build 'smart' constructors for a ToC (that might just map straight to the above, at least right now), so that we can change representations more easily.

Lastly, I think it should look more like
```haskell
toC =
  node toc [
    node refsec [
      node tunits, node tsymb, node tanda
    ] ,
    ...
  ]
```
where `refsec` is the name of the Haskell variable that contains `RefSec $ RefProg intro`.

In other words, all the contents should be created "on its own", and then the toC is an assembly of those, with no information other than structure.

**Comment by tingyuw** [comment]
I don't think ``refsec`` is defined in Drasil. Or do you mean ``refSec`` in ``DPLate``? The type of all section variables is like ``RefSec -> DocSection``; it should take parameters to make it a ``DocSection``. I thought we don't have section variables which don't need parameters, so I was going to create a data which only include the "name" of each section.

**Comment by JacquesCarette** [comment]
I meant that the code would look like
```haskell
refsec :: DocSection -- or whatever the right type is
refsec = RefSec $ RefProg intro
```

**Comment by tingyuw** [comment]
I don't think we have that in the current code. The one we have is the original definition in ``Core.hs``, which is ```newtype RefSec = RefProg Contents```; however, we export this definition, re-define it in ``DocDecl.hs``, and change the type to ``RefSec -> DocSection``, and we export this section constructor for further usage. If we want to use the variables in ``Core.hs``, we have to change the name of ``DocSection`` to avoid ambiguity and export ``DocSection`` from both ``Core.hs`` and ``DocDecl.hs``, but I think we will have to also change all section names (which would cause a lot of confusions). Or should we add the code
```
refsec :: DocSection 
refsec = RefSec $ RefProg intro
```
in each example, and re-use the constructor in ``mkSRS`` and ``toC``?

**Comment by tingyuw** [comment]
Also, I would like to define the ``TableOfContents`` like this:
```haskell 
newtype TableOfContents = ToCProg (Tree DocSection)
```
The ``DocSection`` here is the one defines in ``Core.hs``.

In ``DocDecl.hs``, we re-define some of the sections and change all sections' type to ``DocSection -> Core.DocSection``. This leads to a problem that the type ``DocSection`` of ``TableOfContents`` is inconsistent (It should be type - ``Core.DocSection`` but in ``DocDecl.hs``, it will be ``DocDecl.DocSection``). Furthermore, we only export the ``DocSection`` from ``DocDecl.hs``, therefore, we can't add a ``Core.DocSection`` type in examples. That is, we can't pass the ``DocSection`` parameters (the ``Core.DocSection``) when building examples. 
The solution I can think of is to add something like ``SRSSection`` for ``TableOfContents`` to avoid type inconsistency. In that case, we can build a nested structure directly and don't need the tree structure?

**Comment by tingyuw** [comment]
> The main message is that I'd like each (sub)section to exist as an independent `DocSection`-valued Haskell named entity (whichever version of `DocSection` is the 'right' one for the job).

Each section is now an independent `DocSection`-valued named entity. It's just that neither existing `DocSection` can be used to build the `TableOfContents` tree structure (due to type inconsistency) . I was thinking to add another `DocSection` data, which won't include the definition of the section, but only the 'name' of it.

I'm not sure I understand the 'automatically' here, but it will be generated automatically in the document with other sections. What I'm trying to do is to build a `tocTree` in each example with type `DocSection`, which looks like
```Haskell
tocTree :: Tree DocSection
tocTree = 
  Node TableOfContents [
    Node RefSec [
      Node TUnits [],
      Node TSymb [], ...
```
The table of contents will be built like the other sections in `mkSRS`. `TableOfContents` takes `tocTree` as its argument.
```Haskell
mkSRS = [TableOfContents $ ToCProg tocTree,
   RefSec $ RefProg intro,
    TUnits TUProg, 
    TSymb $ tsymb [TSPurpose, SymbOrder],  ...
```
And to build the section, we will have to iterate over the tree and make each section and its subsections a nested `ItemType`. Although I'm not sure the exact implementation about this part yet, this is my current approach to build the table of contents.

**Comment by JacquesCarette** [comment]
The part with `tocTree` makes sense.  The part with `mkSRS` seems odd. It seems redundant, i.e. repeating the same information.

**Comment by tingyuw** [comment]
`mkSRS` is the original one that exists now, the only change I made is the first line ```TableOfContents $ ToCProg tocTree,``` to build the section.

Besides, I found a problem/technical difficulty with the tree approach. To print the table of contents section nicely in Drasil, we have to use `ItemType: Nested` to print each section (with the subsections) so that the list can be wrapped in `<ul></ul>` like this to make a sub-list
```HTML
<li>
  <a href=#Sec:RefMat>Reference Material</a>
  <ul class="list">
    <li><a href=#Sec:ToU>Table of Units</a></li>
    <li><a href=#Sec:To of Symbols</a></li>
    <li><a href=#Sec:TAbbAcc>Abbreviations and Acronyms</a></li>
  </ul>
 </li>
```
To achieve that, we have to 'separate' each section and its subsection in the tree and make them a `Nested` item. Otherwise, if we just iterated over and print the tree, it will be one flat list (since we can't just add spaces to pretty print the tree in Drasil).
If we still need to do this step with the tree structure, I am thinking to create a nested version of the sections in the core language, since we have already built another flat `DocSection` for it to avoid type inconsistency (the problem I mentioned above), we can make the new `DocSection` nested so we can take advantage of that and build the table of contents section directly. I think it is more straightforward, and if we need nested section structure in the future we don't have to create another tree again.

**Comment by JacquesCarette** [comment]
I was expecting that some of the higher-level routines (like `mkSRS`) would need some adjustment to keep working. "Printing" the new data-structure is going to need a bit of state to work with the new structure, because you need to know what level you're currently at to know whether to output an item or a sub-list.  I also expected some of the current structures would need a bit more adjustment to make things smoother. `Nested` is directly modeled on the fact that our structures were nested, and would remain nested. So all of this might need tweaking (or sometimes, radically rethought).

**Comment by tingyuw** [comment]
I see. So we should stay with the tree approach and adjust the current language and structure for it, right? However, I still finished the code of the nested section last week (the code can be checked on the branch), and I'm thinking to merge the current changes of flatten structure into master and adjust the structure gradually later if that is okay because I'm not sure the scope and how much time will it take for the future changes.

**Comment by JacquesCarette** [comment]
Right.  Yes, I'm fine with merging as is.  Which PR?

**Comment by tingyuw** [comment]
The table of contents remains unchanged as the final implementation does not require any modifications to it. Thus, this issue has been resolved.
